docs/canonical_testing.md

```diff
@@ -0,0 +1,186 @@
+these are encapsulated in the e2e class.
+
+
+## job images
+
+Each class of job will have a companion image that:
+
+* contains all the dependencies required to complete the job class
+* logic to deal with customization that varies per job:
+  - job class customization lives in a well-known folder
+* logic to deal with customization per user:
+  - provider customization (project, creds, etc) lives in another well-known folder
+  - upload customization (same but for uploads) lives in a third well-known folder
+
+For example the e2e images will contain everything to run `hack/e2e.go` as well
+as whatever else remains the same between different jobs in the class.
+
+Each image will expect the customization to live in a particular location. The
+image will
+
+* process this customized information
+* set the customized arguments/environment variables
+* run the job class command included in the image.
+
+## Yaml customization
+
+The way each job will customize the class is through a yaml file. The yaml file
+will specify things that the image needs to customize -- arguments and
+environment variables.
+
+When jobs run locally the top-level runner will duplicate the logic that
+normally happens inside the container: it will know how to process the yaml file
+to set the arguments and variables.
+
+The code to do this inside or outside the container will be the same.
+
+
+# Additional details
+
+##  k8stest
+
+```
+k8stest [ARGS] [JOB ...] [-- JOB_ARGS]
+
+JOB: a test job like kubernetes-build or kubernetes-e2e-gce or job set alias.
+  unit: set of all unit test jobs like verify/build/etc
+  e2e: set of all e2e jobs that run before commit
+  blocking: set of all jobs that block the submit queue
+  critical: set of all jobs that file issues on flakes.
+
+  By default the unit and e2e jobs run.
+
+JOB_ARGS: optionally args to send to every job.
+  Use this to run a subset of tests for example.
+
+ARGS: optional args that control how tests are run.
+  --mode: controls where tests run
+    --mode=local: runs tests locally outside a container (default)
+    --mode=docker: runs tests locally inside a container
+    --mode=k8s: run starts a k8s job to run each tests
+  --config=CONFIG: configuration required to run suites, required for e2e tests
+    defaults to the config section of ~/.k8stest if present
+  --upload=UPLOAD_CONFIG: configuration to upload results, false by default
+    defaults to the upload section of ~/.k8stest if present
+
+  By default we do not build or upload.  and run locally.
+
+CONFIG: a yaml file describing information necessary to deploy a cluster.
+  projects available for use
+  kubectl credentials
+  provider credentials
+
+UPLOAD\_CONFIG: a yaml file describing information necessary to transfer data.
+  gsutil credentials
+  prefix to builds
+  prefix for results
+```
+
+
+## Yaml customization
+
+This will have two keys: `arguments` and `variables`, each of which is a list of
```

Comment:
can you provide a real example? I don't understand how either of these would be used.